nifi-extension-bundles/nifi-network-bundle/nifi-network-processors/src/main/java/org/apache/nifi/processors/network/pcap/SplitPCAP.java

```diff
@@ -0,0 +1,237 @@
+                            throw new ProcessException("PCAP contains an invalid packet. Packet number " + totalPackets + 1 + " is invalid - " + currentPacket.invalidityReason());
+                        }
+
+                        if (currentPcapTotalLength + currentPacket.totalLength() > pcapMaxSize) {
+
+                            templatePcap.packets().addAll(loadedPackets);
+                            FlowFile newFlowFile = session.create(flowFile);
+                            try (final OutputStream out = session.write(newFlowFile)) {
+                                out.write(templatePcap.readBytesFull());
+                                splitFiles.add(newFlowFile);
+                            }
+
+                            loadedPackets.clear();
+                            currentPcapTotalLength = PCAP.PCAP_HEADER_LENGTH;
+                            templatePcap.packets().clear();
+                        }
+
+                        loadedPackets.add(currentPacket);
+                        totalPackets++;
+                        currentPcapTotalLength += currentPacket.totalLength();
+                    }
+
+                    // If there are any packets left over, create a new flowfile.
+                    if (!loadedPackets.isEmpty()) {
+                        templatePcap.packets().addAll(loadedPackets);
+                        FlowFile newFlowFile = session.create(flowFile);
+                        try (final OutputStream out = session.write(newFlowFile)) {
+                            out.write(templatePcap.readBytesFull());
+                            splitFiles.add(newFlowFile);
+                        }
+                    }
+                }
+            });
+        } catch (ProcessException e) {
+            getLogger().error("Failed to split {}", flowFile, e);
+            session.remove(splitFiles);
+            splitFiles.clear();
+            session.putAttribute(flowFile, ERROR_REASON_LABEL, e.getMessage());
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        final String fragmentId = UUID.randomUUID().toString();
+        final String originalFileName = flowFile.getAttribute(CoreAttributes.FILENAME.key());
+        final String originalFileNameWithoutExtension = originalFileName.substring(0, originalFileName.lastIndexOf("."));
+
+        final Map<String, String> attributes = new HashMap<>();
+        attributes.put(FRAGMENT_COUNT, String.valueOf(splitFiles.size()));
+        attributes.put(FRAGMENT_ID, fragmentId);
+        attributes.put(SEGMENT_ORIGINAL_FILENAME, originalFileName);
+
+        IntStream.range(0, splitFiles.size()).forEach(index -> {
+            FlowFile split = splitFiles.get(index);
+            attributes.put(CoreAttributes.FILENAME.key(), originalFileNameWithoutExtension + "-" + index + ".pcap");
+            attributes.put(FRAGMENT_INDEX, Integer.toString(index));
+            session.putAllAttributes(split, attributes);
```

Comment:
This call returns an updated reference, which should be used instead of the previous references.  Is there a reason for not setting these attributes in the main packet processing loop?

Reply:
I'm not sure what you mean with regards to the reference used here, does that mean I should put the call session.transfer inside this loop and call it per 'split' instead of calling it once with 'splitFiles'?

The attributes are all set at the end rather than in the processing loop because it was mentioned that it's a bad idea to call session.putAttribute too many times (link [here](https://github.com/apache/nifi/pull/8691#discussion_r1605163620) and the value for the 'FRAGMENT_COUNT' attribute can only be determined after the main loop completes.

Reply:
> I'm not sure what you mean with regards to the reference used here, does that mean I should put the call session.transfer inside this loop and call it per 'split' instead of calling it once with 'splitFiles'?

`session.putAllAttributes()` returns a new `FlowFile` instance. Instead of creating another `List` of new FlowFile references, it is looks like it would be simpler to call `session.transfer()` for each updated FlowFile reference in this loop.

> 
> The attributes are all set at the end rather than in the processing loop because it was mentioned that it's a bad idea to call session.putAttribute too many times (link [here](https://github.com/apache/nifi/pull/8691#discussion_r1605163620) and the value for the 'FRAGMENT_COUNT' attribute can only be determined after the main loop completes.

Thanks for clarifying, that makes sense in light of the intended behavior.